### PR TITLE
Mostly final version to code review

### PR DESCRIPTION
**Pull request by kkadosa**
Technically a combination of issue 4: Set up the build job to publish the Viatra Gradle Plugin, and issue 7: Provide detailed documentation of the usage of Viatra Gradle Plugin

### CODE REVIEW

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+`input` is the directory of the patterns. Defaults to `src/main/vql`.
+
+Sadly, the output directory can not be configured. It's always `src-gen`.
+
+== Cleaning
+
+This plugin adds the src-gen directory to the targets of the clean task.
+
+== Transitive
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: What does Transitive mean here? Shouldn't it rather be called Summary?

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java [hunk 2]
```diff
@@ -1,46 +1,59 @@
+/*******************************************************************************
+ * Copyright (c) 2020-2020, IncQuery Labs Ltd.
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java]: You can add your name to the copyright headers consistently.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java [hunk 3]
```diff
@@ -35,110 +47,124 @@
...
+            Files.walk(output.getAsFile().toPath()).sorted(Comparator.reverseOrder()).map(Path::toFile)
+                    .forEach(File::delete);
+        } catch (NoSuchFileException ignored) {
+        }
+
+        Project project = getProject();
+        Configuration compileClasspath =
+                project.getConfigurations().getByName(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME);
+
+        //Worst hack of my life: Adds the compile classpath to own classloader via reflection
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: Instead of calling it the worst hack it is more professional to write something as follows (I hope I understood the goal of this change; feel free to correct my mistakes):

// FIXME: In order to access generated metamodel code we have to extend the current classloader via reflection

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java [hunk 4]
```diff
@@ -1,25 +1,52 @@
...
+        //add src-gen to sourceSets.main.java
+        JavaPluginConvention java = project.getConvention().getPlugin(JavaPluginConvention.class);
+        java.getSourceSets().getByName("main").java(conf -> conf.srcDir(SRC_GEN));
+
+        //delete src-gen when cleaning
+        Task clean = tasks.getByName(BasePlugin.CLEAN_TASK_NAME);
+        clean.doFirst(task -> task.getProject().delete(SRC_GEN));
+
+        project.getDependencies().add(JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME,
+                "org.eclipse.viatra:viatra-query-language:2.3.0");
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: We should not hardcode the plugin version here - it will be updated regularly (multiple times a year). Most likely, it should be synchronized with the current release/snapshot versions.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java [hunk 5]
```diff
@@ -1,25 +1,52 @@
...
+
+        RepositoryHandler repos = project.getRepositories();
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: Why is this part necessary? If only because in the VIATRA release is only available here, we should add a TODO comment that this can be removed when we migrate this to the same build infrastructure VIATRA is built upon.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java [hunk 6]
```diff
@@ -1,25 +1,52 @@
...
+        repos.maven(spec -> spec.setUrl("https://repo.eclipse.org/content/groups/releases/"));
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: Why is this release group necessary? I understand the other one, but this generic release group makes little sense to me.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+----
+
+== Usage
+
+For the plugin to work, it needs the metamodel, in the form of compiled java classes, and the fully qualified names of the Package classes.
+
+For the metamodel, add it to the list of dependencies, just as you would any other compile dependency.
+
+For the names, the plugin add an extension, named "vlq", with a method called names, which takes one oe more strings.
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Clarified usage conditions and fixed the typo.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+[source]
+----
+vql {
+	names('org.eclipse.viatra.examples.cps.cyberPhysicalSystem.CyberPhysicalSystemPackage',
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Function renamed to packageClassNames.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+}
+----
+
+In addition, for configuration purposes, the same extension provides other possibilities. The example below shows all possible methods and properties, neither set to their respective default values.
+
+[source]
+----
+vql {
+	names('com.nomagic.uml2.ext.magicdraw.metadata.UMLPackage')
+	updateManifest.set(true)
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Removed those properties.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	generateMatchProcessors.set(false)
+	generateExtensions.set(false)
+	matcherGenerationStrategy.set(org.eclipse.viatra.query.patternlanguage.emf.util.EMFPatternLanguageGeneratorConfig.MatcherGenerationStrategy.SEPARATE_CLASS)
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Separated fully qualified reference into import and short reference.
It could use a string and valueOf(). Should it?

[the remark this answers is not in the document]
Comment (ujhelyiz) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Using the import it looks reasonable. Having a 'compile' error instead of a runtime error is a better approach (I know they are similar).

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 11]
```diff
@@ -47,35 +47,33 @@ plugins {
 
 For the plugin to work, it needs the metamodel, in the form of compiled java classes, and the fully qualified names of the Package classes.
 
-For the metamodel, add it to the list of dependencies, just as you would any other compile dependency.
-
-For the names, the plugin add an extension, named "vlq", with a method called names, which takes one oe more strings.
+* Add the compiled java classes containing the metamodel to your dependencies, just as you would any other dependency.
+* Give the fully qualified names to the plugin. This is done using the `packageClassNames` function of the newly added `vql` extension. It takes any number of strings.
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]:
```suggestion
* A metamodel is registered by the fully qualified class name of the generated EPackage class using the `packageClassNames` function of the newly added `vql` extension. Multiple metamodels can be registered by adding multiple class names as separate strings.
```

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc (hunk 1 above)
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: For example the plugin expects the java plugin to be present, so it applies it itself.
This is hidden from the user, so I wanted to at least write it down somewhere other than the code.
This is kind of like a transitive dependency, that's where the heading came from. I would be happy to use a better title.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java (hunk 2 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java]: My contract explicitly says that I cannot.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java (hunk 3 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: Elaborated the comment to explain the reason of the hack.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc (hunk 1 above)
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Moved the block next to the example.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+	names('com.nomagic.uml2.ext.magicdraw.metadata.UMLPackage')
+	updateManifest.set(true)
+	generateMatchProcessors.set(false)
+	generateExtensions.set(false)
+	matcherGenerationStrategy.set(org.eclipse.viatra.query.patternlanguage.emf.util.EMFPatternLanguageGeneratorConfig.MatcherGenerationStrategy.SEPARATE_CLASS)
+	input.set(project.getLayout().getProjectDirectory().dir('patterns'))
+}
+----
+
+`updateManifest`, `generateMatchProcessors`, `generateExtensions`, and `matcherGenerationStrategy` are the same as normally.
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Now, if the `useEcliseConfig` property is set to true, the plugin will use the config file.

File: viatra-gradle-plugin/documentation/asciidoc/gradle.adoc [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+`input` is the directory of the patterns. Defaults to `src/main/vql`.
+
+Sadly, the output directory can not be configured. It's always `src-gen`.
```
[the remark this answers is not in the document]
Comment (kkadosa) [reply viatra-gradle-plugin/documentation/asciidoc/gradle.adoc]: Created a 'known issues' heading at the end, and moved this there.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 4 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: Can now process a system property named VQL_VERSION.
The plugin will use that version of viarta-query-language as its own dependency, and also when adding it to the user's project.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java [hunk 14]
```diff
@@ -1,46 +1,51 @@
...
-import org.gradle.api.file.DirectoryProperty;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.provider.Property;
 
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
 public class Extension {
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java]: Is it necessary to call this class Extension (e.g. does Gradle expects this name)? If not, I'd rather call it something like VQLGeneratorConfiguration or something similar.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java [hunk 15]
```diff
@@ -12,133 +22,132 @@
...
+        FIXME: Avoid reflective access to own classloader
+        */
+        URLClassLoader cl = (URLClassLoader) this.getClass().getClassLoader();
+        Method add = URLClassLoader.class.getDeclaredMethod("addURL", URL.class);
+        add.setAccessible(true);
+        compileClasspath.resolve().forEach(file -> {
+            try {
+                add.invoke(cl, file.toURI().toURL());
+            } catch (MalformedURLException | IllegalAccessException | InvocationTargetException e) {
+                e.printStackTrace();
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: If this fails, we should not try to fail the build right now instead of trying to do anything else. Furthermore, I am not sure how Gradle usually reports errors, but a printStackTrace is most likely not the recommended way.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java [hunk 16]
```diff
@@ -12,133 +22,132 @@
...
+
+        //Loads metamodels
+        extension.getPackageClassNames().forEach(name -> {
+            try {
+                Class<?> c = cl.loadClass(name);
+                Field instance = c.getDeclaredField("eINSTANCE");
+                EPackage ePackage = (EPackage) instance.get(null);
+                ePackage.getNsURI();
+            } catch (ClassNotFoundException | NoSuchFieldException | IllegalAccessException e) {
+                e.printStackTrace();
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: Similar to the previous commit, we should provide better error handling here. We should report separately that (1) the class was not found or (2) the class is found but was not identified as an EPackage (e.g. does not have any field called eINSTANCE or is of a different type) or have any other issue.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java [hunk 17]
```diff
@@ -12,133 +22,132 @@
...
-import org.gradle.api.tasks.Input;
-import org.gradle.api.tasks.InputDirectory;
-import org.gradle.api.tasks.OutputDirectory;
-import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.*;
 import org.xtext.gradle.builder.XtextGradleBuilder;
 import org.xtext.gradle.protocol.GradleBuildRequest;
-import org.xtext.gradle.protocol.GradleOutputConfig;
 
 public class GenerateVql extends DefaultTask {
```
Comment (ujhelyiz) [inline viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: Formatting question: why does all (or at least most) method declarations in this file start with one or more empty lines?

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 4 above)
Comment (ujhelyiz) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: This seems still hardcoded to me looking at the pull request. If the comment describes the proposed solution, I am fine with that (or if I did miss something, then I sorry for the noise :) ).

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 6 above)
Comment (ujhelyiz) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: Even if we add this group, we should not include the repository URLs as a string in the middle of the code, at least we should extract the constants.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 4 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: Extracted the group and artifact name into constant.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 5 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: This is added to the user's classpath. If they're using this plugin, we know that they'll need it.
It's not necessary, but it's a convenience feature.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java (hunk 6 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/VqlPlugin.java]: I checked, all dependencies of VQL can be reached using Maven Central, so I removed the Eclipse release repo.
Updated the documentation to reflect this.
Extracted VIATRA repo URL into string constant.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java (hunk 14 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/Extension.java]: In Gradle, this kind of class is called an extension.
Renamed to VqlGeneratorExtension.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java (hunk 15 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: The try-catch is inside the loop, so the build continues even if an exception occurs.
The InvocationTargetException cannot happen, as we know that the method is in the interface of the URLClassLoader, and the object is of that class.
The IllegalAccessException cannot happen, as the method was set accessible just before the loop.
The MalformedURLException should not happen, as the configuration is resolved, making all its contents actual files.
At the same time, edge cases occur, and I think whoever happens that to would find a full stacktrace useful.
Added comment explaining this, and a log communicating the file it happened to.

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java (hunk 16 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: Separated the exceptions to their own cathches, and added unique error messages to each of them.
Also added ClassCastException, which is the real test if it's an EPackage (and why the effect-free ePackage.getNsURI() is there).

File: viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java (hunk 17 above)
Comment (kkadosa) [reply viatra-gradle-plugin/src/main/java/org/eclipse/viatra/gradle/querybuilder/GenerateVql.java]: I think you're looking at an older version.
